A3-channel-tracing.md

```diff
@@ -0,0 +1,74 @@
+Channel Tracing
+----
+* Author(s): Noah Eisen (ncteisen@google.com)
+* Approver: markdroth
+* Status: Draft
+* Implemented in: N/A
+* Last updated: 01/02/18
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/WFDj3KeHYTI
+
+## Table of Contents
+
+  * [Abstract](#abstract)
+  * [Background](#background)
+  * [Proposal](#proposal)
+  * [Format of Exported Data](#format-of-exported-data)
+
+## Terminology
+
+This design uses the same terminology discussed in the [channelz](A14-channelz.md) design.
+
+In addition, the following terminology will be used:
+1.  an "trace event" is an interesting thing that happens to a channel. Example include things like creation, address resolution, subchannel creation, connectivity state changes.
+2. "trace data" is all of the data that is contained for one channel. This includes a list of trace events, as well as metadata like the timestamp at which the channel was created.
+3.  a "tracing object" is an in-memory data structure responsible for holding the trace data for a single channel.
+
+## Abstract
+
+The document proposes adding a dedicated, in-memory tracing object for every channel and subchannel. The tracing object will record important events in the life of a channel, like address resolution, subchannel creation, channel state changes, etc etc. The trace data from this tracing object will be made available through a new gRPC API as a JSON formatted string.
```

Comment:
s/JSON/Proto/

We want to pick one serialization and stick with it.  Having both would put undue stress on a client for reading the data.

Reply:
C core does not support proto and has no plans to. Since the new API will be added to the c core, the data has to be exported as a JSON string. The best we can do is ensure the JSON has a deterministic mapping to Proto. This is c core reads service config as JSON.

Reply:
Actually, Java doesn't have direct support of Proto either (and we also have no plans to).   You should still consider using proto to define the messages on the _API_ though, but pass such objects around internally as a struct/class.   

In java, outside of the core library we have a set of services (like health checking and reflection)  which can read the internal datastructures of core.  This allows decoupling of the implementation and the API.

You don't have to do the same approach, but you should consider it, as java team has found it works well and is maintainable.  We are doing the same thing for Channelz, which has both a shim proto service on top of out internal datastructures.